sys/kern/dev_procstat.c

```diff
@@ -0,0 +1,137 @@
+#include <sys/devfs.h>
+#include <sys/vnode.h>
+#include <sys/linker_set.h>
+#include <sys/queue.h>
+#include <sys/proc.h>
+#include <sys/thread.h>
+#include <sys/klog.h>
+#include <sys/malloc.h>
+#include <stdio.h>
+
+/* Implementation of /dev/procstat
+ *
+ * When /dev/procstat is opened it takes a snapshot of info of all existing
+ * processes.
+ *
+ * During read calls to /dev/procstat this info can be read.
+ *
+ * Example:
+ * euid   pid    ppid    pgrp   session  state    td_name  elfpath
+ * 0       1       0       1       1       R       init    /bin/ksh
+ *
+ * XXX: Each instance of /dev/procstat has 4K bytes for process entries.
+ * It will be enough for about 40 processes (if we assume average size of entry
+ * to be 100 characters).
+ *
+ * Until we run more processes than ksh and some tests, this solution don't need
+ * to be more complex.
+ */
+
+#define PSTAT_SIZE 4096
+#define BUF_SIZE (PSTAT_SIZE - sizeof(size_t))
```

Comment:
not so important: maybe it would be better to subtract `sizeof(pstat_t)`, so the macro won't be broken if somebody will add any extra field in the struct.

Reply:
Changed